### PR TITLE
Acme

### PR DESCRIPTION
**Pull request by apmcleod**
make_dataset.py updated.

I added args:
-excerpt-length: Length of the excerpts to include in the dataset (in ms). Default 5s.
-min-notes: Minimum number of notes that must be present for a valid excerpt. Default 10. The script will try up to 10 excerpts until it finds a valid one, or skip the composition. (Sample note, get all notes whose onset is w/in excerpt-length after the sampled note).
-seed: numpy seed. I also shuffle the compositions before iterating through them.
-clean-prop: The proportion of excerpts in the dataset which should be non-degraded. This value is NOT relative to --degradation-dist, but an absolute probability. Degradation-dist will all be scaled down by a factor of (1 - clean-prop).
-splits: Relative proportions of train, test, and validation sets.

I fixed degradation error handling. Degradations that can't be performed just try another degradation, or no degradation (if clean-prop > 0). Otherwise, the composition is skipped (if all degradations fail and clean-prop==0).

This chooses degradations to try in order of how far behind the current distribution is to the desired distribution, so the resulting distribution is as correct as possible, no matter how many compositions must be skipped (and where in the iteration the failed comps are).

Splits are chosen the same way as degradations, so those proportions also match.

I validated and tested some args (ones I didn't are marked with TODOs)

I also changed the organization of the resulting directory. The base dir still contains clean and altered dirs, each with csvs, but not they only contain the excerpts used in the dataset, not the entire piece. Also, only those files contained in the dataset are included. That means that for non-altered excerpts, their input and target csvs are exactly the same file. If someone wants to create a new dataset or a new split, there is no guarantee they will want the same excerpts, or the same distribution of degradations, etc. Plus, the processing is pretty quick anyways.

There is no more metadata directory. Instead, there is a labels.csv file, like this:
```
0,none
1,time_shift
2,onset_shift
...
```

And a metadata.csv file, like this:
```
altered/PianoMidi/bach_850.csv,1,4,clean/PianoMidi/bach_850.csv,train
...
```
The columns are altered_file, has_degraded, degradation_num, clean_file, split

### CODE REVIEW

**Comment by apmcleod** [comment]
Also checked local_midi_dirs is now working, and added recursive option in case local dataset has sub-directories.

Added --clear option to delete input_data and acme directories before creation. I would vote for this being the default functionality, as otherwise leads to strange results sometimes: using datasets not selected because they already exist in input_data, for example.

Made default use no datasets (from Downloads). Previously, there was no way to use none if a user wanted to use only local data. Users explicitly choosing their data should be fine too. We could potentially add an "All" downloader if we want to enable easy selection of all datasets.

**Comment by apmcleod** [comment]
This fixes #31

**Comment by JamesOwers** [comment]
Looks good but we have a bit of a difference in opinion about how to do the degradations:
* I think we should degrade everything, making a clean and degraded version of *all files*; we can then get whatever proportions of degraded vs. non-degraded when performing a split later
* what you've coded here forces the user to choose the degradation level up front, meaning the number of files in ./acme/altered and ./acme/clean do not match

I think my way is better because it is simpler and more flexible:
* we don't need this 'none' degradation
* we can change proportions relative to the task we're interested in / the user can even do train on multiple different splits by selecting a proportion of clean they want and randomly sampling filenames from the ./clean and ./altered directories in that proportion.

Let me know what you think.

**Comment by apmcleod** [comment]
Here was my justification:

I view a dataset as a particular set of data points, with a given split and
proportion already. Other splits or proportions would be considered
different datasets (though they could be drawn from the same distribution
of data points, as we would have here). This is compatible with most large
datasets that I've seen, I think (thinking particularly of maestro and WSJ
at the moment). These have standardized splits set up to make things easily
comparable (which is our goal here, really).

We want the csvs to be excerpts rather than the whole csvs, so we are
losing some flexibility there by which excerpt we take from each piece. We
would almost want to force the user to make different excerpts if they want
a new dataset, rather than making our random excerpts the canonical ones.

For your way, I think we would actually need 1 clean and 9 degraded
versions of each file (1 for each degradation), to support possible future
datasets with different splits (ones without pitch_shift, for example). We
would still need to ensure in the sample that the clean and altered version
of each excerpt are not both used in the same dataset, which adds some
complication in both cases. Plus, we wouldn't want our initial run's
randomly-chosen degradation to be the canonical one for each type (I mean,
for remove_note, we wouldn't want every remove_note with excerpt
bach_850.csv to remove exactly the same note, in every derived dataset).
We'd rather it be truly random I would think.

Given that the dataset can be created quite quickly (~2mins PianoMidi), I
don't think the added flexibility post-creation is worth the potential
complication, compared to a situation where we can just zip the resulting
acme directory and say here are all of the files you need for the acme v1.0
dataset. That, plus the benefit of additional randomness and variation in
chosen excerpt and degradation (even within the same degradation for the
same piece), are the main reasons I prefer it this way.

**Comment by JamesOwers** [comment]
Thanks for super speed response.

Your way makes my life easier in terms of the dataset and dataloader
writing. Your point re *maximum* flexibility (performing all degradations)
is fair, but my way would at least allow a little more information to be
learned from one "dataset" (because you can write a dataloader which
performs a different split on each iteration). But you're right, let's keep
it simple.

Will merge this and write a dataset and loader presuming the
clean/degradation proportion has been done and that no files are in both.

File: make_dataset.py [hunk 1]
```diff
@@ -220,51 +336,166 @@ def parse_args(args_input=None):
                         csv_path=csv_path,
                         read_note_csv_kwargs=read_note_csv_kwargs)
                     for csv_path in tqdm(csv_paths, desc="Cleaning csv data")]
+    np.random.shuffle(compositions) # This is important for join_notes
```
Comment (JamesOwers) [inline make_dataset.py]: Why is this important?

**Comment by apmcleod** [comment]
I agree with you about the download all by default. I think that's fine how you have it now, and my goal was just to enable none selection somehow.

I also agree with you on the naming of metadata and labels, now that you mention it. We could call them {name_of_dataset}.csv (or will this make dataloader more difficult?) or dataset.csv and label_info.csv ?

File: make_dataset.py (hunk 1 above)
Comment (apmcleod) [reply make_dataset.py]: Actually, not specifically for join_notes any more, but it is still important generally.

With the same degradation proportions, 2 runs will perform the degradations in exactly the same order where possible (i.e., first composition: join_notes, 2nd: split_notes, ....). So, if we use the same datasets without shuffling, the same degradation would be performed on the same excerpt every time.

**Comment by JamesOwers** [comment]
Ah, I see! This is as a result of the new way of iterating through and
trying to degrade in a specified order due to current counts. Gotcha.
Before I was making a  random sample of degradations so that would render
this shuffle unnecessary.

**Comment by JamesOwers** [comment]
> I also agree with you on the naming of metadata and labels, now that you mention it. We could call them {name_of_dataset}.csv (or will this make dataloader more difficult?) or dataset.csv and label_info.csv ?

I think metadata is fine, I think I'll just change labels to degradation_ids (I made column degradation_id in the metadata file...)
